django/thunderstore/api/cyberstorm/views/packages.py

```diff
@@ -0,0 +1,434 @@
+    Create base QuerySet for community scoped PackageListAPIViews.
+    """
+
+    return (
+        Package.objects.active()  # type: ignore
+        .select_related("latest", "namespace")
+        .prefetch_related(
+            "community_listings__categories",
+            "community_listings__community",
+        )
+        .exclude(~Q(community_listings__community__pk=community.pk))
+    )
+
+
+def filter_deprecated(
+    show_deprecated: bool,
+    queryset: QuerySet[Package],
+) -> QuerySet[Package]:
+    if show_deprecated:
+        return queryset
+
+    return queryset.exclude(is_deprecated=True)
+
+
+def filter_nsfw(
+    show_nsfw: bool,
+    queryset: QuerySet[Package],
+) -> QuerySet[Package]:
+    if show_nsfw:
+        return queryset
+
+    return queryset.exclude(community_listings__has_nsfw_content=True)
+
+
+def filter_in_categories(
+    category_ids: List[int],
+    queryset: QuerySet[Package],
+) -> QuerySet[Package]:
+    """
+    Include only packages belonging to specific categories.
+
+    Multiple categories are OR-joined, i.e. if category_ids contain A
+    and B, packages belonging to either will be returned.
+    """
+    if not category_ids:
+        return queryset
+
+    return queryset.exclude(
+        ~Q(community_listings__categories__id__in=category_ids),
+    )
+
+
+def filter_not_in_categories(
+    category_ids: List[int],
+    queryset: QuerySet[Package],
+) -> QuerySet[Package]:
+    """
+    Exclude packages belonging to specific categories.
+
+    Multiple categories are OR-joined, i.e. if category_ids contain A
+    and B, packages belonging to either will be rejected.
+    """
+    if not category_ids:
+        return queryset
+
+    return queryset.exclude(
+        community_listings__categories__id__in=category_ids,
+    )
+
+
+def filter_by_section(
+    section_uuid: Optional[str],
+    queryset: QuerySet[Package],
+) -> QuerySet[Package]:
+    """
+    PackageListingSections can be used as shortcut for multiple
+    category filters.
+    """
+    if not section_uuid:
+        return queryset
+
+    try:
+        section = PackageListingSection.objects.prefetch_related(
+            "require_categories",
+            "exclude_categories",
+        ).get(uuid=section_uuid)
+    except PackageListingSection.DoesNotExist:
+        required = []
+        excluded = []
+    else:
+        required = section.require_categories.values_list("pk", flat=True)
+        excluded = section.exclude_categories.values_list("pk", flat=True)
+
+    queryset = filter_in_categories(required, queryset)
+    return filter_not_in_categories(excluded, queryset)
+
+
+def filter_by_query(
+    query: Optional[str],
+    queryset: QuerySet[Package],
+) -> QuerySet[Package]:
+    """
+    Filter packages by free text search.
+    """
+    if not query:
+        return queryset
+
+    search_fields = ("name", "owner__name", "latest__description")
+    icontains_query = Q()
+    parts = [x for x in query.split(" ") if x]
```

Comment:
You could use a generator here instead of a list, just swap `[]` wrapping to `()`. Generators are a sane default for transformations if you don't need to cache the result (like here, as we're only looping over it once). Saves memory and a for loop, sometimes in significant amounts depending on how much data you're working with.

Obviously not a blocker